Review Charm for Listing

kelkawi-a opened the pull request:
PR for reviewing charm code as defined in: https://discourse.charmhub.io/t/reviewing-charms/11698.

dimaqq reviewed (commented):
### First pass review


Very high-level question, out of curiousity:

What's the motivation of charming Airtable specifically.

For example, there seem to be open-source alternatives...


Re Airtable specifically, I think there's a Helm Chart for that, how does this Charm compare to that?

kelkawi-a commented on src/scripts/pod-sweeper.sh:
```diff
@@ -0,0 +1,69 @@
+# https://github.com/airbytehq/airbyte-platform/blob/main/charts/airbyte-pod-sweeper/templates/configmap.yaml
+# TODO(kelkawi-a): Move this to Airbyte ROCK
+
+get_job_pods() {
+    # echo "Running kubectl command to get job pods..."
+    kubectl -n "${JOB_KUBE_NAMESPACE}" -L airbyte -l airbyte=job-pod \
+        get pods \
+        -o=jsonpath='{range .items[*]} {.metadata.name} {.status.phase} {.status.conditions[0].lastTransitionTime} {.status.startTime}{"\n"}{end}'
+}
+
+delete_pod() {
+    printf "From status '%s' since '%s', " "$2" "$3"
+    echo "$1" | grep -v "STATUS" | awk '{print $1}' | xargs --no-run-if-empty kubectl -n "${JOB_KUBE_NAMESPACE}" delete pod
+}
```
Airbyte itself creates k8s pods for different sync jobs. This script is meant to periodically clean up job pods which have exceeded a certain ttl value. I've added a note to the README to reflect that Airbyte creates these pods. 

Scaling with this charm has not been tested yet, but this might be modified depending on the outcome of such tests.

kelkawi-a commented on charmcraft.yaml:
```diff
@@ -0,0 +1,467 @@
+      description: Name of logs storage bucket.
+      default: "airbyte-dev-logs"
+      type: string
+
+    logs-ttl:
+      description: |
+        Number of days until logs are purged from object storage.
+      default: 30
+      type: int
+
+    storage-bucket-state:
+      description: Name of state storage bucket.
+      default: "airbyte-state-storage"
+      type: string
+
+    storage-bucket-activity-payload:
+      description: Name of activity payload storage bucket.
+      default: "airbyte-payload-storage"
+      type: string
+
+    storage-bucket-workload-output:
+      description: Name of workload output storage bucket.
+      default: "airbyte-state-storage"
+      type: string
+
+    ##### Miscellaneous config #####
+    pod-running-ttl-minutes:
+      description: Number of minutes until a running job pod is removed.
+      default: 240
+      type: int
+
+    pod-successful-ttl-minutes:
+      description: Number of minutes until a successful job pod is removed.
+      default: 30
+      type: int
+
+    pod-unsuccessful-ttl-minutes:
+      description: Number of minutes until an unsuccessful job pod is removed.
+      default: 1440
+      type: int
```
Airbyte by design is a complex application which exposes many configurations to limit resource usage on the k8s cluster. By default, the user can deploy the charm and leave all default configurations as is. These configurations are enabled to give flexibility to the user to add kube tolerations, node selectors and annotations. They are also here to have parity with the official Airbyte Helm chart as discussed with the Airbyte team.

dimaqq replied:
Well, this flies in the face of "The Model Driven Operator Manifesto"

https://juju.is/model-driven-operations-manifesto

> High-level config
> Operators should abstract complex application configuration. An operator should offer only high-level config to capture admin intent and reduce the burden of application specific domain expertise required to run the app.

P.S. I don't know the context behind this charm or who the airbyte team is, if possible please link some discussion.

My 2c: it's not a blocker, but exposing this many configuration knobs taxes administrators in the field.

kelkawi-a commented on tests/unit/test_charm.py:
```diff
@@ -0,0 +1,449 @@
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+
+
+"""Charm unit tests."""
+
+# pylint:disable=protected-access,too-many-public-methods
+
+import logging
+from unittest import TestCase, mock
```
Is this blocking? We currently use `unittest.TestCase` across all our team's charms, and this would require refactoring all the tests.

kelkawi-a commented on src/charm.py:
```diff
@@ -0,0 +1,318 @@
+            container = self.unit.get_container(container_name)
+            valid_pebble_plan = self._validate_pebble_plan(container, container_name)
+            logger.info(f"validating pebble plan for {container_name}")
+            if not valid_pebble_plan:
+                logger.debug(f"failed to validate pebble plan for {container_name}, attempting creation again")
+                all_valid_plans = False
+                self._update(event)
+                continue
+
+            logger.info(f"performing up check for {container_name}")
+            check = container.get_check("up")
+            if check.status != CheckStatus.UP:
+                logger.error(f"check failed for {container_name}")
```
I tend to prefer more verbose code, given that there are multiple if checks being performed inside the for-loop.

kelkawi-a commented on src/charm.py:
```diff
@@ -0,0 +1,318 @@
+            if not valid_pebble_plan:
+                logger.debug(f"failed to validate pebble plan for {container_name}, attempting creation again")
+                all_valid_plans = False
+                self._update(event)
```
Good catch, i've updated this to only be called once if the `all_valid_plans` evaluates to `False`.

kelkawi-a commented on tox.ini:
```diff
@@ -0,0 +1,115 @@
+    isort {[vars]src_path} {[vars]tests_path}
+    black {[vars]src_path} {[vars]tests_path}
+
+[testenv:lint]
```
I wasn't able to re-create this, perhaps a mismatch in tox versions?

```
git clone git@github.com:canonical/airbyte-k8s-operator.git
cd airbyte-k8s-operator
python -m venv venv
source venv/bin/activate
pip install -r dev-requirements.txt
tox
```

dimaqq replied:
The issue here is that the tool is linting not only the charm code, but also the virtualenv that tox has created.

My host Python is 3.12 and `SafeConfigParser` was removed by that version. I would imagine that pflake8 has some kind of work-around included.

In short, it's not an issue for the "conventional" way of running things, but the command could be improved to scan only some files. This approach admittedly has its own problems.

If it works for you locally and in CI, keep it.

kelkawi-a commented on src/literals.py:
```diff
@@ -0,0 +1,64 @@
+    "CONFIG_ROOT": "/configs",
+    "CONFIGS_DATABASE_MINIMUM_FLYWAY_MIGRATION_VERSION": "0.35.15.001",
+    "JOBS_DATABASE_MINIMUM_FLYWAY_MIGRATION_VERSION": "0.29.15.001",
+    "MICRONAUT_ENVIRONMENTS": "control-plane",
+    "WORKERS_MICRONAUT_ENVIRONMENTS": "control-plane",
+    "CRON_MICRONAUT_ENVIRONMENTS": "control-plane",
+    "WORKLOAD_API_HOST": "localhost",
+    "MICROMETER_METRICS_ENABLED": "false",
+    "LAUNCHER_MICRONAUT_ENVIRONMENTS": "control-plane,oss",
```
Micronaut is used in Airbyte's [tech stack](https://docs.airbyte.com/understanding-airbyte/tech-stack), these environment variables are needed to configure it to the OSS deployment.

kelkawi-a commented on charmcraft.yaml:
```diff
@@ -0,0 +1,467 @@
+# See https://juju.is/docs/sdk/metadata-reference for a checklist and guidance.
+
+# Your workload’s containers.
+containers:
+  airbyte-api-server:
+    resource: airbyte-api-server
+  airbyte-bootloader:
+    resource: airbyte-bootloader
+  airbyte-connector-builder-server:
+    resource: airbyte-connector-builder-server
+  airbyte-cron:
+    resource: airbyte-cron
+  airbyte-pod-sweeper:
+    resource: airbyte-pod-sweeper
+  airbyte-server:
+    resource: airbyte-server
+  airbyte-workers:
+    resource: airbyte-workers
```
I did not find much guidance on the approach of using services vs. containers for a multi-container application. It might be possible to do this using services, but I am not sure how that would affect the charm operationally (e.g. observing logs).

We have not yet designed the scaling aspect of the Airbyte charms, but yes, the current assumption is that each juju unit/pod would contain these containers.

dimaqq replied:
Wrt. scaling, I figure it's fair game to take inspiration from the existing Helm chart.

kelkawi-a commented on src/state.py:
```diff
@@ -0,0 +1,66 @@
+        self.__dict__["_app"] = app
+        self.__dict__["_get_relation"] = get_relation
+
+    def __setattr__(self, name, value):
```
This is actually being used across the charm for storing relation information (e.g. database and object storage credentials). An example of this can be seen under `relations/postgresql.py`.

dimaqq replied:
Thank you, found them. I'll leave comments there.

kelkawi-a commented on src/state.py:
```diff
@@ -0,0 +1,66 @@
+import json
+
+
+class State:
```
Noted the typed wrapper as a future improvement on this wrapper class.

kelkawi-a commented on tests/unit/test_charm.py:
```diff
@@ -0,0 +1,449 @@
+                    "SYNC_JOB_RETRIES_PARTIAL_FAILURES_MAX_TOTAL": 20,
+                    "TEMPORAL_HISTORY_RETENTION_IN_DAYS": 30,
+                    "TEMPORAL_HOST": "temporal-k8s:7233",
+                    "TEMPORAL_WORKER_PORTS": "9001,9002,9003,9004,9005,9006,9007,9008,9009,9010,9011,9012,9013,9014,9015,9016,9017,9018,9019,9020,9021,9022,9023,9024,9025,9026,9027,9028,9029,9030",
```
These are listed as an environment variable that is used by Airbyte. Reference [here](https://github.com/airbytehq/airbyte-platform/blob/275832edb013879ec475430fca7cef9c04854124/charts/airbyte/templates/env-configmap.yaml#L140).

kelkawi-a commented:
> ### First pass review
> Very high-level question, out of curiousity:
> 
> What's the motivation of charming Airtable specifically.
> 
> For example, there seem to be open-source alternatives...
> 
> Re Airtable specifically, I think there's a Helm Chart for that, how does this Charm compare to that?

Thanks for the thorough review! The use of Airbyte is a business decision as its architecture matches what we are looking for in an ETL system, including the use of Temporal for durable execution of synchronization workflows.

This charm mostly has parity with the Helm chart in terms of the configs it exposes and the functionality it provides through the default deployment.

I've addressed most of the review comments in #16

dimaqq commented on src/relations/minio.py:
```diff
@@ -0,0 +1,137 @@
+                storage_data["secure"],
+            )
+
+            self.charm._state.minio = {
+                **storage_data,
+                "endpoint": endpoint,
+            }
+            self.charm._update(event)
```
Python doesn't enforce the distinction between public and private member/method names.

Then again this very PR creates this member and method.

Would it make sense to make them "public", i.e. `self.charm.state` and `self.charm.update()`?

dimaqq commented on tests/unit/test_structured_config.py:
```diff
@@ -0,0 +1,90 @@
+
+
+def test_memory_related_values(_harness) -> None:
+    """Test specific parameters for each field."""
```
Same docstring appearing in 3 different tests is not very useful.

dimaqq reviewed (commented):
Here's another batch of comments.

Sorry that I'm unable to review the whole thing. I'm multi-tasking and the charm is kinda large.

dimaqq commented on charmcraft.yaml:
```diff
@@ -0,0 +1,467 @@
+    limit: 1
+    optional: true
+
+provides:
+  airbyte-server:
+    interface: airbyte-server
```
Discussion on naming style: https://matrix.to/#/!xdClnUGkurzjxqiQcN:ubuntu.com/$oWHOH6C6yg_Ni0XpIq8X7zgGfLDgas9BEcie6xBIyMY?via=ubuntu.com&via=matrix.org&via=matrix.debian.social

dimaqq reviewed (changes requested):
Please read https://juju.is/docs/sdk/holistic-vs-delta-charms

This applies to the airbyte-server relation event handling, where:
- if the relation is set up before the unit is ready, no data is published yet (as far as I can tell)
- charm relies on udpate-status event to detect that it's ready and publish relation data

The update status events happen every 5 minutes by default, however, the period is often set longer, possibly much longer in production. Which means that it may take a while for the application and relations to settle.

Additionally I'm not very happy to see `_provide_server_status` a private-ish method being called from another module.

Ideally, the charm would be event driven, where relation data may be ultimately published on some container's pebble ready event. The implementation would then likely look like a "holistic" charm.

Alternatively, you could set up a background process in a container, that checks stuff and issues a custom pebble notice when things change.

You could also chose to bite the bullet and keep the current architecture, even if it's a little hard to understand.

As a reviewer, my requirement is that there's an explicit choice.

kelkawi-a commented:
> Please read https://juju.is/docs/sdk/holistic-vs-delta-charms
> 
> This applies to the airbyte-server relation event handling, where:
> 
> * if the relation is set up before the unit is ready, no data is published yet (as far as I can tell)
> * charm relies on udpate-status event to detect that it's ready and publish relation data
> 
> The update status events happen every 5 minutes by default, however, the period is often set longer, possibly much longer in production. Which means that it may take a while for the application and relations to settle.
> 
> Additionally I'm not very happy to see `_provide_server_status` a private-ish method being called from another module.
> 
> Ideally, the charm would be event driven, where relation data may be ultimately published on some container's pebble ready event. The implementation would then likely look like a "holistic" charm.
> 
> Alternatively, you could set up a background process in a container, that checks stuff and issues a custom pebble notice when things change.
> 
> You could also chose to bite the bullet and keep the current architecture, even if it's a little hard to understand.
> 
> As a reviewer, my requirement is that there's an explicit choice.

This architecture is being used on other charms developed by our team (e.g. [Temporal](https://github.com/canonical/temporal-k8s-operator), [Trino](https://github.com/canonical/trino-k8s-operator)), so i'd like to keep it as is for uniformity. There also doesn't seem to be a consensus around naming style for interfaces, some charms use kebab-case (e.g. nginx-ingress-integrator) for their interface.

The `airbyte-server` container is the charm container which handles Airbyte server logic, while the `airbyte-server` relation is the endpoint that the [Airbyte UI charm](https://github.com/canonical/airbyte-ui-k8s-operator) relates to to fetch information about the Airbyte server.

dimaqq commented:
Minor comment to the above:
- Temporal charm does work in the same way indeed: https://github.com/canonical/temporal-k8s-operator/blob/45f4a36e5e0ec6a1190a14004103fa4e4da7f387/src/charm.py#L262-L287
- Trino charm does not: https://github.com/canonical/trino-k8s-operator/blob/4be540bebe5c28e77656607f51dd3fe064cf5acd/src/charm.py#L187-L195 (on_update_status does not trigger relation data update, it relies on event deferral instead)

dimaqq commented:
We've discussed this in Charm Tech, and found the following:
- our general responsibility is to advise on good practices
- in rare cases (e.g. security or approach unlikely to work at all), we may block a charm
- the choice of architecture is the decision of your team
- improvement in the way each charm team structures their charms is viewed as an ongoing process that can't a specific charm review cannot be waiting for, rather it ought to be done out of band, through training, tooling or office hours

In summary, go ahead with your current approach and fixes staged in #16

dimaqq commented on README.md:
```diff
@@ -0,0 +1,103 @@
+This charm is still in active development. Please see the
+[Juju SDK docs](https://juju.is/docs/sdk) for guidelines on enhancements to this
+charm following best practice guidelines, and
+[CONTRIBUTING.md](./CONTRIBUTING.md) for developer guidance.
```
Please use a link with full hostname, so that the link works when this file is displayed on charmhub.

dimaqq commented on charmcraft.yaml:
```diff
@@ -0,0 +1,467 @@
+
+name: airbyte-k8s
+type: charm
+title: Airbyte Server
```
I would prefer if the CharmHub name and the README title matched.

Today. it's a little off.
![Screenshot 2024-07-26 at 17 02 15](https://github.com/user-attachments/assets/72182bf7-f653-4347-a672-186dafce5f15)

dimaqq commented on tests/unit/test_structured_config.py:
```diff
@@ -0,0 +1,90 @@
+
+
+@pytest.fixture
+def _harness():
```
Minor:

there are special rules for pytest fixtures.
because of that, there isn't really a "private" method notation.

```suggestion
def harness():
```

P.S. a fixture with the same name overrides a fixture from an outer scope, so even if there is a `harness` fixture somewhere, a custom name is not called for.

dimaqq commented on .github/workflows/publish_charm.yaml:
```diff
@@ -0,0 +1,13 @@
+name: Publish to edge
+
+on:
+  push:
+    branches:
+      - main
```
I must be a little confused... are tests run before the charm is published to the edge channel?

kelkawi-a replied:
The tests are run on the PR, once the tests pass, this workflow runs when the PR is merged with main to push the changes to the `edge` channel on Charmhub.

dimaqq commented on .github/workflows/test.yaml:
```diff
@@ -0,0 +1,9 @@
+name: Tests
+
+on:
+  pull_request:
```
Unless that's implemented in some other way, I would recommend to also run tests on push to main.

The reason for that is if you've got 2 PR's, each may independently pass the tests, but then both are merged, the aggregate may still fail.

kelkawi-a replied:
This was the case previously with the [operator workflows](https://github.com/canonical/operator-workflows) developed by the IS Devops team, but was changed in favor of only running the tests on the PR. PRs will only be merged one at a time to ensure such a case does not happen.
